src/main/java/org/junitpioneer/jupiter/CartesianEnumSource.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2016-2021 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.jupiter;
+
+import static java.util.Arrays.stream;
+import static java.util.stream.Collectors.toSet;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Repeatable;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.BiPredicate;
+import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
+
+import org.junit.jupiter.params.provider.ArgumentsSource;
+import org.junit.platform.commons.PreconditionViolationException;
+
+/**
+ * {@code @CartesianEnumSource} is an argument source for constants of a
+ * specified {@linkplain #value Enum}.
+ *
+ * <p>The enum constants will be provided as an argument source to
+ * the corresponding parameter of the annotated {@code @CartesianProductTest} method.
+ *
+ * <p>The set of enum constants can be restricted via the {@link #names} and
+ * {@link #mode} attributes.
+ *
+ * <p>This annotation is {@link Repeatable}. You should declare one
+ * {@code @CartesianEnumSource} per parameter.
+ * </p>
+ *
+ * @see CartesianProductTest
+ *
+ * @since 1.2.1
+ */
+@Target({ ElementType.ANNOTATION_TYPE, ElementType.METHOD })
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Repeatable(CartesianEnumSource.CartesianEnumSources.class)
+@ArgumentsSource(CartesianEnumArgumentsProvider.class)
+public @interface CartesianEnumSource {
+
+	/**
+	 * The enum type that serves as the source of the enum constants.
+	 *
+	 * @see #names
+	 * @see #mode
+	 */
+	Class<? extends Enum<?>> value();
+
+	/**
+	 * The names of enum constants to provide, or regular expressions to select
+	 * the names of enum constants to provide.
+	 *
+	 * <p>If no names or regular expressions are specified, all enum constants
+	 * declared in the specified {@linkplain #value enum type} will be provided.
+	 *
+	 * <p>The {@link #mode} determines how the names are interpreted.
+	 *
+	 * @see #value
+	 * @see #mode
+	 */
+	String[] names() default {};
+
+	/**
+	 * The enum constant selection mode.
+	 *
+	 * <p>Defaults to {@link CartesianEnumSource.Mode#INCLUDE INCLUDE}.
+	 *
+	 * @see CartesianEnumSource.Mode#INCLUDE
+	 * @see CartesianEnumSource.Mode#EXCLUDE
+	 * @see CartesianEnumSource.Mode#MATCH_ALL
+	 * @see CartesianEnumSource.Mode#MATCH_ANY
+	 * @see #names
+	 */
+	CartesianEnumSource.Mode mode() default CartesianEnumSource.Mode.INCLUDE;
```

Comment:
Just for the record, I don't really like this way to structure the API - having one "general purpose" container (`names`) with an only half-way matching name (how do `names` relate to `MATCH_ALL`?) and then `mode` that defines how to process it, feels off to me. I'd prefer an API with straightforward attributes `String[] include()`, etc. (although that opens the can of what-if-multiple-attributes-are-set can of worms).

Anyway, since we want this annotation to mimic Jupiter's and it does the same thing, this is perfectly fine.